### PR TITLE
Rename ImmutableArray to ArraySeq

### PR DESCRIPTION
**Pull request by lrytz**
Fixes https://github.com/scala/bug/issues/10836

### CODE REVIEW

**Comment by lrytz** [comment]
(this also needs a rebase on the partest in-source)

**Comment by NthPortal** [comment]
I think we may need another collection - a `ConSeq` - for this

**Comment by lrytz** [comment]
rebased on the insource-partest pr for now to test, but unfortunately that created 100 jobs on jenkins, as there's no `[ci: last-only]` here...

**Comment by lrytz** [comment]
Tests look good, but the `specialized` category failed because I need to build a new `instrumented.jar`. I did that recently, so no big deal.

**Comment by lrytz** [comment]
Added another commit that renames `WrappedArray` to `ArraySeq`. So basically there's `mutable.ArraySeq` and `immutable.ArraySeq`. @szeiger suggested this in https://github.com/scala/bug/issues/10836. I'm not 100% sure it's a good idea, because creating a `mutable.ArraySeq` is not always explicit, due to `Predef.wrapXArray`. It might be too confusing. Let's discuss!

**Comment by lrytz** [comment]
Also, if we name it `mutable.ArraySeq` instead of `mutable.WrappedArray`, we basically change the semantics of `mutable.ArraySeq` from 2.12 to 2.13. Users of 2.12 `ArraySeq` would have to use `ArraySeq.untagged`. We could add a migration warning.

The change of semantics is also there in current 2.13.x, `mutable.ArraySeq` is a deprecated alias for `WrappedArray`. So users get a deprecation warning. If we keep this, we should also add a migration warning.

File: src/library/scala/collection/mutable/package.scala [hunk 1]
```diff
@@ -2,10 +2,6 @@ package scala.collection
 
 
 package object mutable {
-  @deprecated("Use WrappedArray instead of ArraySeq; it can represent both, boxed and unboxed arrays", "2.13.0")
-  type ArraySeq[X] = WrappedArray[X]
-  @deprecated("Use WrappedArray instead of ArraySeq; it can represent both, boxed and unboxed arrays", "2.13.0")
-  val ArraySeq = WrappedArray.untagged
```
Comment (sjrd) [inline src/library/scala/collection/mutable/package.scala:8]: I think we need the deprecated aliases in the other direction, if we're doing this change.

Comment (lrytz) [reply src/library/scala/collection/mutable/package.scala:8]: Yes, thanks for pointing that out. Though I think we should drop the second commit, keep `mutable.WrappedArray`. Waiting for opinions on that still :-)

**Comment by julienrf** [comment]
> Users of 2.12 `ArraySeq` would have to use `ArraySeq.untagged`. We could add a migration warning.

I agree. And we can easily define a rewrite rule replacing `ArraySeq.empty[X]` to `ArraySeq.untagged.empty[X]`, and same for all the other operations that now take an implicit `ClassTag` at

**Comment by lrytz** [comment]
Gah, this also needs a new scalacheck release.. I can do that through a `publishLocal`'d scala version i guess.

There's also a binary incompatibility in the new macroAnnots tests.

```
[test] [error] - macroAnnot/test
[test] [error]   - macroAnnot/compile:compileIncremental failed: java.lang.NoSuchMethodError: scala.runtime.ScalaRunTime$.wrapRefArray([Ljava/lang/Object;)Lscala/collection/immutable/ImmutableArray;
```

@adriaanm are there any pre-compiled binaries involved here? Otherwise it must be that code is being compiled with the starr compiler, but then executed with locker or quick. Just running `macroAnnot/test` locally worked for me, so I'm not sure.

**Comment by lrytz** [comment]
It turns out just rebuilding scalacheck with a local Scala version is not enough, because the scalacheck artifact depends on the Scala library it was built with. So that Scala version needs to be on maven... We can either
  1. Release another `2.13.0-M4-pre-abcd123` to maven (with the rename) and the build scalacheck using that
  2. In-source scalacheck
  3. other ideas?

I vote for 2.

**Comment by SethTisue** [comment]
👍 to in-sourcing

**Comment by lrytz** [comment]
Added a commit that in-sources scalacheck. TODO: update IntelliJ templates.

File: src/library/scala/Array.scala [hunk 2]
```diff
@@ -494,11 +494,11 @@ object Array {
    *  @return  sequence wrapped in a [[scala.Some]], if `x` is an Array, otherwise `None`
    */
   def unapplySeq[T](x: Array[T]): Option[IndexedSeq[T]] =
-    if (x == null) None else Some(ImmutableArray.unsafeWrapArray[T](x))
+    if (x == null) None else Some(ArraySeq.unsafeWrapArray[T](x))
     // !!! the null check should to be necessary, but without it 2241 fails. Seems to be a bug
     // in pattern matcher.  @PP: I noted in #4364 I think the behavior is correct.
-    // Is ImmutableArray safe here? In 2.12 we used to call .toIndexedSeq which copied the array
-    // instead of wrapping it in a WrappedArray but it appears unnecessary.
+    // Is ArraySeq safe here? In 2.12 we used to call .toIndexedSeq which copied the array
```
Comment (szeiger) [inline src/library/scala/Array.scala:500]: It turns out it is not. We need to make a defensive copy. Varargs use `immutable.Seq` so we can't just return a `mutable.ArraySeq` but the Seq can leak out of a pattern match.

**Review by szeiger — APPROVED** [review]

**Comment by sjrd** [comment]
FYI, it turns out Scala.js also depends on this PR to be merged, if it does and what portions of it do :-s I had not realized that before.

I've already updated after #6620.

**Comment by lrytz** [comment]
I'll merge this as soon as it goes green.

**Pull request merged**